lib/bolt/executor.rb

```diff
@@ -347,6 +357,41 @@ def run_plan(scope, plan, params)
       plan.call_by_name_with_scope(scope, params, true)
     end
 
+    def handle_event(event)
+      case event[:type]
+      when :node_result
+        # Give the result a chance to return back to the function
+        # If we resume the fiber before the result makes it there, this will
+        # spin forever
+        #
+        # TODO: Is there a way to avoid this?
+        sleep(1)
+        @thread_completed = true
+      end
+    end
+
+    def round_robin(skein)
+      subscribe(self, [:node_result])
+      results = Array.new(skein.length)
+      @in_parallel = true
+      until skein.empty?
+        @thread_completed = false
+        skein.each do |yarn|
+          # Check if stored a value
+          if yarn.alive?
+            yarn.resume
+          else
+            results[yarn.index] = yarn.value
+            skein.delete(yarn)
+          end
+        end
+        sleep(1) until @thread_completed
```

Comment:
```suggestion
        skein.each do |yarn|
          # Check if stored a value
          if yarn.alive?
            yarn.resume
          else
            results[yarn.index] = yarn.value
            skein.delete(yarn)
          end
        end
```

I'm not sure what `@thread_completed` is doing, and it seems like it can result in things running slowly _and_ cause the plan to get stuck in an infinite loop. Since `@thread_completed` can only be set to `true` by the executor receiving a `node_result` event, the block would need to end with a function call that publishes that event. For example, the following would cause an infinite loop:

```
plan bolt () {
  $numbers = [1, 2, 3]
  $result = parallelize($numbers) |$number| {
    $number * 2
  }
  return $result
}
```

I also ran into an infinite loop a few other times with blocks that ended with a `run_*` function call.

Reply:
Hm - we need @thread_completed to tell when we're waiting for several tasks to complete so we don't thrash while waiting for several threads to finish. I think I can yield an argument from functions that yield the Fiber and then check for the argument in the loop, so that anything that doesn't yield back `round_robin` won't wait for the node_result call.

Reply:
This is exactly the kind of thing I wanted to catch now! Thanks Tom!!